README.md
# express-maintenance-mode

[![code coverage](https://img.shields.io/codecov/c/github/dissfall/express-maintenance-mode.svg)](https://codecov.io/gh/dissfall/express-maintenance-mode)
[![code style](https://img.shields.io/badge/code_style-XO-5ed9c7.svg)](https://github.com/sindresorhus/xo)
[![styled with prettier](https://img.shields.io/badge/styled_with-prettier-ff69b4.svg)](https://github.com/prettier/prettier)
[![made with lass](https://img.shields.io/badge/made_with-lass-95CC28.svg)](https://lass.js.org)
[![license](https://img.shields.io/github/license/dissfall/express-maintenance-mode.svg)](LICENSE)
[![npm downloads](https://img.shields.io/npm/dt/express-maintenance-mode.svg)](https://npm.im/express-maintenance-mode)

> Express middleware that allows you to put the server(s) in maintenance mode


## Table of Contents

* [Motivation](#motivation)
* [How it works](#how-it-works)
* [Install](#install)
* [Usage](#usage)
  * [Redis example](#redis-example)
* [API](#api)
* [Contributors](#contributors)
* [License](#license)


## Motivation

I developed this module because I needed an easy way to transfer all servers behind the load balancer
to maintenance mode with one API request. That is why I made methods for getting and setting external maintenance status.
In my production environment, I use these techniques to keep the service status in sync with Redis.
This way, all servers in the load balancing group are aware of the status update within a minute.


## How it works

Module provides simple api to control maintenance state.
Access to all middlewares below can be controlled by maintenance middleware


## Install

[npm][]:

```sh
npm install express-maintenance-mode
```

[yarn][]:

```sh
yarn add express-maintenance-mode
```


## Usage

```typescript
import { ExpressMaintenanceMode } from 'express-maintenance-mode';

const maintenance = new ExpressMaintenanceMode<MaintenanceResponseBody>({
  maintenancePath: '/maintenance', // Path to control maintenance state
  apiBasePath: '/api', // Base path of your API
  accessKey: 'changeme', // Access key for maintenance endpoint. Works without authorization if not provided
  getExternalMaintenanceState: () => {
    // Optional
    // Your method to get external state
  },
  setExternalMaintenanceState: () => {
    // Optional
    // Your method to set external state
  },
  localMaintenanceStateTTL: 6000 // Lifetime of local maintenance state, until it be synced with external state
});

// Optional
interface MaintenanceResponseBody {
  message: string;
}

// App example

app.use(bodyParser.json());
app.use(maintenance.middleware)
// ...
app.use(yourGreatAPIRouter)


```

### Redis example

```typescript
import {MaintenanceState} from './index';

const getExternalMaintenanceState = async (): Promise<MaintenanceState<MaintenanceResponseBody>> => {
  return yourRedisDAO.get<MaintenanceState<MaintenanceResponseBody>>('maintenance');
}
  
const setExternalState = async (maintenanceState: MaintenanceState<MaintenanceResponseBody>) => {
  yourRedicDAO.save('maintenance', maintenanceState);
}
```


## API

To control maintenance status three methods available:

GET - to get maintenance status
POST - to set server in maintenance mode
DELETE - to set server in regular mode

Using POST method you can set response status code and body

Request body:

```json
{
  "statusCode": 503,
  "body": {
    "message": "Server in maintenance mode"
  }
}

```


## Contributors

| Name                |
| ------------------- |
| **George Lykuanov** |


## License

[Apache-2.0](LICENSE) © George Lykuanov


##

[npm]: https://www.npmjs.com/

[yarn]: https://yarnpkg.com/


CHANGELOG.md
# Changelog


## 1.1.0 (2021-05-01)

### Features

* implement middleware ([293d9ff](https://github.com/dissfall/express-maintenance-mode/commit/293d9ff8c341b1047486707931d985ba67b35d18))


test/api.test.ts
import test from 'ava';
import { spy, match, SinonSpy } from 'sinon';
import { Request, Response } from 'express';
import { ExpressMaintenance } from '../src';

type MockResponse = { json: SinonSpy; status: SinonSpy } & Partial<Response>;

test('Middleware should be function', (t) => {
  const maintenance = new ExpressMaintenance();
  t.is(typeof maintenance.middleware, 'function');
});

test('GET /middleware should return server maintenance state', async (t) => {
  const maintenance = new ExpressMaintenance();
  const mockRequest = {
    path: '/maintenance',
    method: 'GET',
    query: {}
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request,
    mockResponse as Response,
    spy()
  );
  t.true(mockResponse.status.calledWith(200));
  t.true(
    mockResponse.json.calledWithMatch(
      match.hasNested('message', match(/default/))
    )
  );
});

test('POST /middleware should turn server to maintenance mode', async (t) => {
  const maintenance = new ExpressMaintenance();
  const mockRequest = {
    path: '/maintenance',
    method: 'POST',
    query: {}
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request,
    mockResponse as Response,
    spy()
  );
  t.true(mockResponse.status.calledWith(200));
  t.true(
    mockResponse.json.calledWithMatch(
      match.hasNested('message', match(/maintenance/))
    )
  );
});

test('Application API should not be available when server in maintenance mode', async (t) => {
  const maintenance = new ExpressMaintenance();
  const mockRequest = {
    path: '/maintenance',
    method: 'POST',
    query: {}
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request,
    // @ts-expect-error
    {
      status: spy(),
      json: spy()
    } as Response,
    spy()
  );
  await maintenance.middleware(
    {
      path: '/api/something',
      url: '/api/something',
      method: 'GET',
      query: {}
    } as Request,
    mockResponse as Response,
    spy()
  );
  t.is(mockResponse.status.getCall(-1).firstArg, 503);
});

test('POST /middleware with body should turn server to maintenance mode and set response options', async (t) => {
  const maintenance = new ExpressMaintenance<{ message: string }>();
  const mockRequest = {
    path: '/maintenance',
    url: '/maintenance',
    method: 'POST',
    query: {},
    body: {
      body: { message: 'Server in maintenance mode now' },
      statusCode: 503
    }
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request,
    mockResponse as Response,
    spy()
  );
  t.true(mockResponse.status.calledWith(200));
  t.true(
    mockResponse.json.calledWithMatch(
      match.hasNested('message', match(/maintenance/))
    )
  );
  const context = maintenance.getContext();
  t.is(context.body.message, mockRequest.body.body.message);
  t.is(context.statusCode, mockRequest.body.statusCode);
});

test('DELETE /middleware should turn server in regular mode', async (t) => {
  const maintenance = new ExpressMaintenance<{ message: string }>();
  const mockRequest = {
    path: '/maintenance',
    url: '/maintenance',
    method: 'DELETE',
    query: {}
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request,
    mockResponse as Response,
    spy()
  );
  t.true(mockResponse.status.calledWith(200));
  t.true(
    mockResponse.json.calledWithMatch(
      match.hasNested('message', match(/default/))
    )
  );
});

test('Attempt to execute API with others methods should be failed', async (t) => {
  const maintenance = new ExpressMaintenance<{ message: string }>();
  const mockRequest = {
    path: '/maintenance',
    url: '/maintenance',
    method: 'PUT',
    query: {}
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request,
    mockResponse as Response,
    spy()
  );
  t.true(mockResponse.status.calledWith(405));
  t.regex(mockResponse.json.getCall(-1).firstArg.message, /not/);
});


test/auth.test.ts
import test from 'ava';
import { spy, SinonSpy } from 'sinon';
import { ExpressMaintenance } from '../src';
import { Request, Response } from 'express';

type MockResponse = { json: SinonSpy; status: SinonSpy } & Partial<Response>;

test('API should not be available without authorization if it enabled', async (t) => {
  const maintenance = new ExpressMaintenance({
    accessKey: 'jive_belarus'
  });
  const mockRequest = {
    path: '/maintenance',
    method: 'POST',
    query: {}
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request,
    mockResponse as Response,
    spy()
  );
  t.is(mockResponse.status.getCall(-1).firstArg, 401);
});

test('Attempt to access API with invalid key should be failed', async (t) => {
  const maintenance = new ExpressMaintenance({
    accessKey: 'jive_belarus'
  });
  const mockRequest = {
    path: '/maintenance',
    method: 'POST',
    query: { accessKey: 'gestachaishe' }
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request & { query: { accessKey: string } },
    mockResponse as Response,
    spy()
  );
  t.is(mockResponse.status.getCall(-1).firstArg, 401);
});

test('Attempt to access API with valid key should be succeed', async (t) => {
  const maintenance = new ExpressMaintenance({
    accessKey: 'jive_belarus'
  });
  const mockRequest = {
    path: '/maintenance',
    method: 'POST',
    query: { accessKey: 'jive_belarus' }
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request & { query: { accessKey: string } },
    mockResponse as Response,
    spy()
  );
  t.is(mockResponse.status.getCall(-1).firstArg, 200);
});


test/external.test.ts
import test from 'ava';
import { SinonSpy, spy } from 'sinon';
import {
  ExpressMaintenance,
  GetExternalMaintenanceStateFunction,
  MaintenanceState,
  ServerMode,
  SetExternalMaintenanceStateFunction
} from '../src';
import { Request, Response } from 'express';

type MockResponse = { json: SinonSpy; status: SinonSpy } & Partial<Response>;
type MaintenanceResponseBody = { string: string };

const getExternalState: GetExternalMaintenanceStateFunction<MaintenanceResponseBody> = (): MaintenanceState<MaintenanceResponseBody> => {
  return {
    currentServerMode: ServerMode.maintenance
  };
};

const getExternalStateAsync: GetExternalMaintenanceStateFunction<MaintenanceResponseBody> = async (): Promise<
  MaintenanceState<MaintenanceResponseBody>
> => {
  return {
    currentServerMode: ServerMode.maintenance
  };
};

const setExternalState: SetExternalMaintenanceStateFunction<MaintenanceResponseBody> = spy(
  // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
  (maintenanceState: MaintenanceState<MaintenanceResponseBody>) => {}
);

test('Middleware should be in sync with external state if external state function provided', async (t) => {
  const maintenance = new ExpressMaintenance({
    getExternalMaintenanceState: getExternalState
  });
  const mockRequest = {
    url: '/maintenance',
    path: '/maintenance',
    method: 'GET',
    query: {}
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request,
    mockResponse as Response,
    spy()
  );
  t.regex(mockResponse.json.getCall(-1).firstArg.message, /maintenance/);
});

test('Middleware should be in sync with external state if external state function provided. Even if it async!', async (t) => {
  const maintenance = new ExpressMaintenance({
    getExternalMaintenanceState: getExternalStateAsync
  });
  const mockRequest = {
    url: '/maintenance',
    path: '/maintenance',
    method: 'GET',
    query: {}
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request,
    mockResponse as Response,
    spy()
  );
  t.regex(mockResponse.json.getCall(-1).firstArg.message, /maintenance/);
});

test('Middleware should be update external state if update function provided', async (t) => {
  const maintenance = new ExpressMaintenance({
    setExternalMaintenanceState: setExternalState.bind(this)
  });
  const mockRequest = {
    url: '/maintenance',
    path: '/maintenance',
    method: 'POST',
    query: {}
  };
  const mockResponse: MockResponse = {
    status: spy(),
    json: spy()
  };
  await maintenance.middleware(
    mockRequest as Request,
    mockResponse as Response,
    spy()
  );
  t.regex(
    (setExternalState as SinonSpy).getCall(-1).firstArg.currentServerMode,
    /maintenance/
  );
});


src/index.ts
import { NextFunction, Request, Response, RequestHandler } from 'express';

export class ExpressMaintenance<
  MaintenanceResponseBody extends Record<string, any>
> {
  private readonly maintenancePath: string = '/maintenance';
  private readonly apiBasePath: string = '/api';
  private readonly accessKey?: string;
  private readonly localMaintenanceStateTTL: number = 60000;

  private readonly getExternalMaintenanceState: GetExternalMaintenanceStateFunction<MaintenanceResponseBody>;
  private readonly setExternalMaintenanceState: SetExternalMaintenanceStateFunction<MaintenanceResponseBody>;

  private lastStateUpdateTimestamp: Date;
  private currentServerMode: ServerMode = ServerMode.default;
  private maintenanceResponseOptions: MaintenanceResponseOptions<MaintenanceResponseBody> = {
    statusCode: 503
  };

  constructor(options?: ExpressMaintenanceOptions<MaintenanceResponseBody>) {
    if (options) {
      this.maintenancePath = options.maintenancePath ?? this.maintenancePath;
      this.apiBasePath = options.apiBasePath ?? this.apiBasePath;
      this.accessKey = options.accessKey ?? this.accessKey;
      this.localMaintenanceStateTTL =
        options.localMaintenanceStateTTL ?? this.localMaintenanceStateTTL;
      this.getExternalMaintenanceState =
        options.getExternalMaintenanceState ?? this.getExternalMaintenanceState;
      this.setExternalMaintenanceState =
        options.setExternalMaintenanceState ?? this.setExternalMaintenanceState;
    }

    this.lastStateUpdateTimestamp = new Date();
  }

  public get middleware(): RequestHandler {
    return async (request: Request, response: Response, next: NextFunction) => {
      await this.updateLocalMaintenanceState();

      if (this.isServerInMaintenanceMode() && this.isApiRequest(request)) {
        response.status(this.maintenanceResponseOptions.statusCode);
        return response.json(this.maintenanceResponseOptions.body);
      }

      if (this.isMaintenanceRequest(request)) {
        const { accessKey } = request.query;
        if (this.accessKey && accessKey !== this.accessKey) {
          response.status(401);
          return response.json({
            message: 'You not authorized to perform this action'
          });
        }

        switch (request.method) {
          case 'GET':
            response.status(200);
            return response.json({
              message: `Server in ${this.currentServerMode} mode now`
            });
          case 'POST':
            this.currentServerMode = ServerMode.maintenance;
            this.maintenanceResponseOptions.statusCode =
              request.body?.statusCode ??
              this.maintenanceResponseOptions?.statusCode;
            this.maintenanceResponseOptions.body =
              request.body?.body ?? this.maintenanceResponseOptions?.body;
            if (this.setExternalMaintenanceState) {
              await this.setExternalMaintenanceState(
                this.getLocalMaintenanceState()
              );
            }

            break;
          case 'DELETE':
            this.currentServerMode = ServerMode.default;
            if (this.setExternalMaintenanceState) {
              await this.setExternalMaintenanceState(
                this.getLocalMaintenanceState()
              );
            }

            break;
          default:
            response.status(405);
            return response.json({
              message: `${request.method} is not allowed for this endpoint`
            });
        }

        response.status(200);
        return response.json({
          message: `Server in ${this.currentServerMode} mode now`,
          maintenanceResponseOptions: this.maintenanceResponseOptions
        });
      }

      next();
    };
  }

  private async updateLocalMaintenanceState(): Promise<void> {
    if (
      this.getExternalMaintenanceState &&
      this.isItTimeToUpdateLocalMaintenanceState()
    ) {
      const mayBePromise = this.getExternalMaintenanceState();
      const externalMaintenanceState: MaintenanceState<MaintenanceResponseBody> =
        mayBePromise[Symbol.toStringTag] === 'Promise'
          ? await mayBePromise
          : (mayBePromise as MaintenanceState<MaintenanceResponseBody>);
      if (!externalMaintenanceState) {
        return;
      }

      const {
        currentServerMode,
        maintenanceResponseOptions
      }: MaintenanceState<MaintenanceResponseBody> = externalMaintenanceState;
      this.currentServerMode = currentServerMode;
      this.maintenanceResponseOptions = maintenanceResponseOptions;
      this.lastStateUpdateTimestamp = new Date();
    }
  }

  private isItTimeToUpdateLocalMaintenanceState(): boolean {
    if (process.env.NODE_ENV === 'test') {
      return true;
    }

    return (
      this.lastStateUpdateTimestamp.getTime() + this.localMaintenanceStateTTL <
      Date.now()
    );
  }

  private isServerInMaintenanceMode(): boolean {
    return this.currentServerMode === ServerMode.maintenance;
  }

  private isApiRequest(request: Request): boolean {
    return request.path.includes(this.apiBasePath);
  }

  private isMaintenanceRequest(request: Request): boolean {
    const urlRegExp = new RegExp(`${this.maintenancePath}$`);
    return Boolean(urlRegExp.test(request.path));
  }

  private getLocalMaintenanceState(): MaintenanceState<MaintenanceResponseBody> {
    return {
      currentServerMode: this.currentServerMode,
      maintenanceResponseOptions: this.maintenanceResponseOptions
    };
  }

  public getContext(): ExpressMaintenanceOptions<MaintenanceResponseBody> &
    MaintenanceResponseOptions<MaintenanceResponseBody> {
    return {
      maintenancePath: this.maintenancePath,
      apiBasePath: this.apiBasePath,
      accessKey: this.accessKey,
      localMaintenanceStateTTL: this.localMaintenanceStateTTL,
      getExternalMaintenanceState: this.getExternalMaintenanceState,
      setExternalMaintenanceState: this.setExternalMaintenanceState,
      ...this.maintenanceResponseOptions
    };
  }
}

export type GetExternalMaintenanceStateFunction<
  MaintenanceResponseBody extends Record<string, any>
> = () =>
  | MaintenanceState<MaintenanceResponseBody>
  | Promise<MaintenanceState<MaintenanceResponseBody>>;

export type SetExternalMaintenanceStateFunction<
  MaintenanceResponseBody extends Record<string, any>
> = (
  maintenanceState: MaintenanceState<MaintenanceResponseBody>
) => void | Promise<void>;

export interface MaintenanceState<
  MaintenanceResponseBody extends Record<string, any>
> {
  currentServerMode?: ServerMode;
  maintenanceResponseOptions?: MaintenanceResponseOptions<MaintenanceResponseBody>;
}

export enum ServerMode {
  default = 'default',
  maintenance = 'maintenance'
}

export interface MaintenanceResponseOptions<
  MaintenanceResponseBody extends Record<string, any>
> {
  statusCode?: number;
  body?: MaintenanceResponseBody;
}

export interface ExpressMaintenanceOptions<
  MaintenanceResponseBody extends Record<string, any>
> {
  maintenancePath?: string;
  apiBasePath?: string;
  accessKey?: string;
  localMaintenanceStateTTL?: number;
  getExternalMaintenanceState?: GetExternalMaintenanceStateFunction<MaintenanceResponseBody>;
  setExternalMaintenanceState?: SetExternalMaintenanceStateFunction<MaintenanceResponseBody>;
}
